base/inference.jl

```diff
@@ -1476,6 +1512,25 @@ function Pair_name()
     return _Pair_name
 end
 
+_typename(a) = Union{}
+_typename(a::Vararg) = Any
+_typename(a::TypeVar) = Any
+_typename(a::DataType) = Const(a.name)
+function _typename(a::Union)
+    ta = _typename(a.a)
+    tb = _typename(a.b)
+    ta == tb ? tb : (ta === Any || tb == Any) ? Any : Union{}
```

Comment:
No, I don't this it should since since `Const(T.name) != Const(T.name)`. `tb === Any is fine though`.